### PR TITLE
Louvain algorithm

### PR DESCRIPTION
**Pull request by jalving**
Here is a simple implementation of the Louvain fast-unfolding algorithm I whipped up.  It has been giving valid communities for my problems.

The algorithm could use some speed-up enhancements.  Notably, I'm using the LightGraphs.modularity function to calculate modularity change in my loop as opposed to calculating the change directly.

Please feel free to comment or make changes.

### CODE REVIEW

**Comment by simonschoelly** [comment]
I think your pr kind of went unnoticed, as there is not much activity on this package at the moment. I will try to have a look at your pr soon.

**Comment by jalving** [comment]
No rush, I've been using community detection methods in my research and figured this is the best place to share the work.

**Comment by VPetukhov** [comment]
Hi @simonschoelly , do you by chance have any updates on this?

**Comment by storopoli** [comment]
I did some benchmarking and the algorithm is performing very well

```julia
using BenchmarkTools
G = erdos_renyi(100, 0.1)

@benchmark community_detection_louvain(G)
@benchmark community_detection_bethe(G)
```

### Benchmark for Bethe

```
BenchmarkTools.Trial: 
  memory estimate:  523.06 KiB
  allocs estimate:  266
  --------------
  minimum time:     1.485 ms (0.00% GC)
  median time:      1.766 ms (0.00% GC)
  mean time:        1.943 ms (3.15% GC)
  maximum time:     46.315 ms (0.00% GC)
  --------------
  samples:          2572
  evals/sample:     1
```

### Benchmark for Louvain

```
BenchmarkTools.Trial: 
  memory estimate:  181.54 MiB
  allocs estimate:  77601
  --------------
  minimum time:     337.115 ms (7.82% GC)
  median time:      352.989 ms (7.50% GC)
  mean time:        355.509 ms (7.61% GC)
  maximum time:     382.444 ms (5.86% GC)
  --------------
  samples:          15
  evals/sample:     1
```

I think it's time to do a pull request and make it to the codebase

File: src/CommunityDetection.jl [hunk 1]
```diff
@@ -107,4 +109,130 @@ function community_detection_bethe(g::AbstractGraph, k::Int=-1; kmax::Int=15)
...
+        ind = rhs[i,:].nzind[1]     #get the index of the vertex
+        labels[i] = ind
+    end
+
+    return labels
+end
+
+function louvain_make_pass(graph::AbstractGraph,communities::Vector{Int},tol::Float64)
+    comm_list = copy(communities)
+    current_modularity = modularity(graph,comm_list)
```
Comment (storopoli) [inline src/CommunityDetection.jl:166]: Is it possible to use a memoing function or to have the `current_modularity` cached? I think this is the major bottleneck in performance. Did you profiled the code to inspect what are the major bottlenecks?

Comment (jalving) [reply src/CommunityDetection.jl:166]: Thanks for taking a look at the algorithm.  I never did any profiling but it shouldn't be too bad to check the bottlenecks. I'll take a look at improving the benchmark. Maybe we can merge once we're happy with the performance?

Comment (storopoli) [reply src/CommunityDetection.jl:166]: Try [`Memoize.jl`](https://github.com/JuliaCollections/Memoize.jl)

**Comment by gdalle** [comment]
Closed in favor of #4

**Pull request closed without merging**